### PR TITLE
fix: stateful use of IGitModule

### PR DESCRIPTION
**Pull request by RussKie**

### CODE REVIEW

File: GitCommands/Remote/GitRemoteController.cs [hunk 1]
```diff
@@ -69,12 +68,12 @@ public class GitRemoteController : IGitRemoteController
...
-        private readonly IGitModule _module;
+        private readonly Func<IGitModule> _getModule;
 
 
-        public GitRemoteController(IGitModule module)
+        public GitRemoteController(Func<IGitModule> getModule)
         {
-            _module = module;
+            _getModule = getModule;
             Remotes = new BindingList<GitRemote>();
```
Comment (jbialobr) [inline GitCommands/Remote/GitRemoteController.cs]: It seems risky to have a state (Remotes) captured from module that can change overtime.

File: GitUI/CommandsDialogs/FormPull.cs [hunk 2]
```diff
@@ -117,7 +117,7 @@ public FormPull(GitUICommands aCommands, string defaultRemoteBranch, string defa
 
             if (aCommands != null)
             {
-                _gitRemoteController = new GitRemoteController(Module);
+                _gitRemoteController = new GitRemoteController(() => Module);
```
Comment (jbialobr) [inline GitUI/CommandsDialogs/FormPull.cs]: These are modal dialogs and it is not possible to change the GitModule during they lifespan. Perhaps there is a need for apply different names to stateless controllers and stateful ones.

Comment (RussKie) [reply GitUI/CommandsDialogs/FormPull.cs]: Indeed the dialog are unlikely to be affected by statefulness of IGitModule. However this way we ensure a consistency across API, which is IMO significantly more important.
I have been going through the codebase for a number of months and yet I have overlooked the statefulness of IGitModule, neither this omission was picked during reviews. If it wasn't for @gerhardol we'd probably get a few bug reports. 
So a chance that another contributor may not be able to understand the implications of statefulness is probably quite high.

File: GitCommands/Remote/GitRemoteController.cs (hunk 1 above)
Comment (RussKie) [reply GitCommands/Remote/GitRemoteController.cs]: this is needs to be refactored out, as it makes the controller stateful...
i'll do as a separate piece of work

File: GitUI/CommandsDialogs/FormPull.cs (hunk 2 above)
Comment (jbialobr) [reply GitUI/CommandsDialogs/FormPull.cs]: The problem I see is when it comes to async calls. GitModule being a function can be easily overlooked to be passed as a captured state to an async call (especially for async await constructs).

**Comment by jbialobr** [comment]
> Do you see any problems with GitGpgController.cs

Yes, I do. See the line https://github.com/gitextensions/gitextensions/pull/3835/files#diff-da9708e394cb1db5b508ef9e7a8cc3a7R154

GetTagVerificationMessage captures GitModule to late, it has to be captured before the task has started.
Otherwise there is no guarantee the command is run against the same GitModule as preceding calls.

I prefer the React.js way where developers deal with the current snapshot of the app state. Transitions from one state to another are handled at generic level. I would see here some analogous solution where GitModule should be a form's and a controller's immutable member. Then the consumers don't have to worry about the GitModules state and simply use it. Usually a form or controller has a single point of receiving GitModule and many points of accessing GitModule, thus it is better to tackle the problem at the receiving point, preferably at the generic level.

**Comment by jbialobr** [comment]
>That's a valid concern as well.... Feels like damned if we do, and damned
if we don't :)
GitModule is crying for a refactor, and one day I hope to get to it.

I feel like refactoring GitModule can't help here. 
I'd rather introduce some encapsulation class to handle async calls and continuations, to provide a consistent environment to them as well as recognizing when the environment got stale and there is no point in performing the continuation. There is an extensions method that does it for Forms disposals:
https://github.com/gitextensions/gitextensions/blob/9485424618886b978e1a05162f3f1274bc036f10/GitUI/GitUIExtensions.cs#L303
Another example is the Lazy class which encapsulates lazy value initialization.

**Comment by RussKie** [comment]
> I feel like refactoring GitModule can't help here

I respectfully disagree, separating the state from behavior will help.
Here's a very rough take on this:
https://github.com/RussKie/gitextensions/tree/refactor_move_state_out_gitmodule

Excluded all unit tests and plugins, the core compiles and runs.

Ideally I would like to take it further - move the remaining state from
GitModule (I think there are few more properties) as well as sever the
injection of GitModuleState into GitModule.
Once this is done, it will be much easier to reason about the GitModule - I
expect quite a few API simply and/or go (e.g. static vs instance). It will
also be possible to ensure that only GitModule (i.e. behaviors) are
injected as opposed to GitModuleState.

I also suspect that if we make GitModuleState a struct, the problem of
thread-safety goes away since it will be passed by value.
I haven't thought it through though, not sure whether there will be any
performance implications.


If we consider this proposal as a possible way forward I would want to
release 2.51 before this goes in. Maybe target the next major release
altogether (e.g. v3.0)



On 20 November 2017 at 05:10, Janusz Białobrzewski <notifications@github.com

**Comment by gerhardol** [comment]
I do not feel I have much to contribute to this change (and other similar, I look at many changes but cannot say I review them).  So slightly offtopic.

> If we consider this proposal as a possible way forward I would want to release 2.51 before this goes in. Maybe target the next major release altogether (e.g. v3.0)

This sounds like a plan to me. Anywhere the contents or release date is discussed other than the milestone? I guess  some of the tagged issues will not be corrected, but there are a few open PRs too (I have at least one more cleanup depending on those too, if the open issues are not merged I should maybe make a modified PR).

The contents of a 3.0 release should probably be defined. A cleaned up interface in Browse is a good starting point. (Not important for me personally, but a GUI is a selling point.)

**Comment by jbialobr** [comment]
>I respectfully disagree...

I am sorry but I don't see how your changes address my concerns. Perhaps you try to solve the problem of GitModule not being thread-safe. Could you please explain what you try to achieve?

**Comment by RussKie** [comment]
> I am sorry but I don't see how your changes address my concerns. 
They don't, not in the immediate PR anyway.
I will rework this PR to address the immediate thread-safety concerns.

> Perhaps you try to solve the problem of GitModule not being thread-safe. Could you please explain what you try to achieve?

At the moment GitModule represents a big non-SOLID furry ball of code, a mix of state and various behaviors (some of which are totally unrelated to the rest). My crude example was meant to show how we may attempt to resolve this strategically.
A lot of code requires access to GitModule to get the state information (i.e. working directory), and a lot of code depends on GitModule for its behaviors. Thread safety (or better to say, "non-safety") concerns relate to the state rather than the behavior.
If the API are split, the behaviors can be safely injected via constructors and the state can/should only be passed as input parameters. 
This should address the concerns you've raised across the whole codebase. But, as I said above, this doesn't address it for the current PR.

As a side note, while working on my separation POC, I've come across a few instances of safe use of GitModule across threads, and a few instances of unsafe use - these will need to be addressed separately.

**Comment by RussKie** [comment]
> but cannot say I review them

You're welcome to start ;)

> Anywhere the contents or release date is discussed other than the milestone? 

No. At the moment I am stretched thin between family and work. I hope to have some free time over the Christmas break, perhaps then we can start fleshing out what next major release should look like.
This is a separate discussion though.

**Comment by RussKie** [comment]
@jbialobr  I've update the PR, there was only one instance of unsafe passing of GitModule to a background thread.
The example you've shown in GitGpgController isn't part of this PR, so I will get it addressed it #3835.

I will also have a sweep of the code to try to ensure similar implementations are fixed (I've seen a few of them). I'll do it as a separate PR.

**Pull request closed without merging**